pallets/parachain-staking/src/tests.rs

```diff
@@ -711,129 +682,225 @@ fn leave_candidates_event_emits_correctly() {
 		.with_candidates(vec![(1, 10)])
 		.build()
 		.execute_with(|| {
-			assert_ok!(Stake::leave_candidates(Origin::signed(1), 1u32));
-			assert_eq!(
-				last_event(),
-				MetaEvent::Stake(Event::CollatorScheduledExit(1, 1, 3))
+			assert_ok!(Stake::schedule_leave_candidates(Origin::signed(1), 1u32));
+			assert_last_event!(MetaEvent::Stake(Event::CandidateScheduledExit(1, 1, 3)));
+		});
+}
+
+#[test]
+fn leave_candidates_removes_candidate_from_candidate_pool() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 10)])
+		.with_candidates(vec![(1, 10)])
+		.build()
+		.execute_with(|| {
+			assert_eq!(Stake::candidate_pool().0.len(), 1);
+			assert_ok!(Stake::schedule_leave_candidates(Origin::signed(1), 1u32));
+			assert!(Stake::candidate_pool().0.is_empty());
+		});
+}
+
+#[test]
+fn cannot_leave_candidates_if_not_candidate() {
+	ExtBuilder::default().build().execute_with(|| {
+		assert_noop!(
+			Stake::schedule_leave_candidates(Origin::signed(1), 1u32),
+			Error::<Test>::CandidateDNE
+		);
+	});
+}
+
+#[test]
+fn cannot_leave_candidates_if_already_leaving_candidates() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 10)])
+		.with_candidates(vec![(1, 10)])
+		.build()
+		.execute_with(|| {
+			assert_ok!(Stake::schedule_leave_candidates(Origin::signed(1), 1u32));
+			assert_noop!(
+				Stake::schedule_leave_candidates(Origin::signed(1), 1u32),
+				Error::<Test>::CandidateAlreadyLeaving
 			);
 		});
 }
 
 #[test]
-fn leave_candidates_unreserves_balance() {
+fn insufficient_leave_candidates_weight_hint_fails() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 20), (2, 20), (3, 20), (4, 20), (5, 20)])
+		.with_candidates(vec![(1, 20), (2, 20), (3, 20), (4, 20), (5, 20)])
+		.build()
+		.execute_with(|| {
+			for i in 1..6 {
+				assert_noop!(
+					Stake::schedule_leave_candidates(Origin::signed(i), 4u32),
+					Error::<Test>::TooLowCandidateCountToLeaveCandidates
+				);
+			}
+		});
+}
+
+#[test]
+fn sufficient_leave_candidates_weight_hint_succeeds() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 20), (2, 20), (3, 20), (4, 20), (5, 20)])
+		.with_candidates(vec![(1, 20), (2, 20), (3, 20), (4, 20), (5, 20)])
+		.build()
+		.execute_with(|| {
+			let mut count = 5u32;
+			for i in 1..6 {
+				assert_ok!(Stake::schedule_leave_candidates(Origin::signed(i), count));
+				count -= 1u32;
+			}
+		});
+}
+
+// EXECUTE LEAVE CANDIDATES
+
+#[test]
+fn execute_leave_candidates_emits_event() {
+	ExtBuilder::default()
+		.with_balances(vec![(1, 10)])
+		.with_candidates(vec![(1, 10)])
+		.build()
+		.execute_with(|| {
+			assert_ok!(Stake::schedule_leave_candidates(Origin::signed(1), 1u32));
+			roll_to(30);
```

Comment:
Why do we have roll_to(30)? dont we schedule it to be 2 the delay in the mock? And if so, why dont we test for a delay of 2?

Reply:
It could be `roll_to(10)` and I'll change it. There were a lot of tests so I was probably lazy at the beginning and didn't check the default round length (which is 5 and assumes it was 10 or 15).

Reply:
The delay is 2 rounds, each round is 5 blocks in length.

The benchmarking code has `roll_to(round: u32)` and I was considering moving that everywhere else...

Reply:
oh nice roll_to(10) looks fine. Moving it to a specific round is perhaps more clear, but this is good